hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStorageContainerManagerFacade.java

```diff
@@ -396,22 +445,90 @@ private void initializeSCMDB() {
       }
     } catch (IOException e) {
       LOG.error("Exception encountered while getting SCM DB.");
+    } finally {
+      isSyncDataFromSCMRunning.compareAndSet(true, false);
     }
   }
 
   public void updateReconSCMDBWithNewSnapshot() throws IOException {
-    DBCheckpoint dbSnapshot = scmServiceProvider.getSCMDBSnapshot();
-    if (dbSnapshot != null && dbSnapshot.getCheckpointLocation() != null) {
-      LOG.info("Got new checkpoint from SCM : " +
-          dbSnapshot.getCheckpointLocation());
+    if (isSyncDataFromSCMRunning.compareAndSet(false, true)) {
+      DBCheckpoint dbSnapshot = scmServiceProvider.getSCMDBSnapshot();
+      if (dbSnapshot != null && dbSnapshot.getCheckpointLocation() != null) {
+        LOG.info("Got new checkpoint from SCM : " +
+            dbSnapshot.getCheckpointLocation());
+        try {
+          initializeNewRdbStore(dbSnapshot.getCheckpointLocation().toFile());
+        } catch (IOException e) {
+          LOG.error("Unable to refresh Recon SCM DB Snapshot. ", e);
+        }
+      } else {
+        LOG.error("Null snapshot location got from SCM.");
+      }
+    } else {
+      LOG.warn("SCM DB sync is already running.");
+    }
+  }
+
+  public boolean syncWithSCMContainerInfo()
+      throws IOException {
+    if (isSyncDataFromSCMRunning.compareAndSet(false, true)) {
       try {
-        initializeNewRdbStore(dbSnapshot.getCheckpointLocation().toFile());
+        List<ContainerInfo> containers = containerManager.getContainers();
+
+        long containerCount = scmServiceProvider.getContainerCount(
+            HddsProtos.LifeCycleState.CLOSED);
+        // Assumption of size of 1 container info object here is 1 MB
+        long containersMetaDataTotalRpcRespSizeMB =
+            CONTAINER_METADATA_SIZE * containerCount;
+        double numOfApiCalls =
+            Double.valueOf(containersMetaDataTotalRpcRespSizeMB) /
+                Double.valueOf(DEFAULT_RPC_SIZE);
+        long totalApiCalls = Math.round(Math.ceil(numOfApiCalls));
+        long containerCountPerCall = containersMetaDataTotalRpcRespSizeMB <=
+            DEFAULT_RPC_SIZE ? containerCount : DEFAULT_RPC_SIZE;
+        if (containerCount > 0) {
+          for (int startContainerId = 1; startContainerId <= totalApiCalls;
```

Comment:
startContainerId should start with 0 index excluding, to get next set of list, and next startIndex should be last queried containerId, this need fix at SCM and recon.